upsilonconf/io.py

```diff
@@ -225,26 +225,109 @@ def _get_io_function(path: Path, write: bool = False):
     raise ValueError(f"unknown config file extension: '{ext}'")
 
 
-def load(path: Union[Path, str]) -> Configuration:
+def __replace_in_keys(
+        dictionary: Union[Dict[str, Any], list, Any],
+        s: str,
+        r: str) -> Union[Dict[str, Any], list, Any]:
+    """
+    Take a list or dictionary and replace all occurrencies of `s` in any of its
+    keys with `r`.
+
+    This function is called recursively.
+
+    Modelled after the Stack Overflow answer by Farhan Haider:
+    https://stackoverflow.com/questions/21650850/pyyaml-replace-dash-in-keys-with-underscore#answer-55986782
+
+    Parameters
+    ----------
+    dictionary : dict|list
+        The list or dictionary to be modified.
+    s : str
+        The string to be replaced.
+    r : str
+        The replacement string.
+
+    Returns
+    -------
+        The object of the same type with all strings `s` replaced by `r`.
+    """
+    # For lists perform this method on every object
+    if isinstance(dictionary, list):
+        return [__replace_in_keys(item, s, r) for item in dictionary]
+
+    # For dictionaries traverse all the keys and replace 's' with 'r'
+    elif isinstance(dictionary, dict):
+        final_dict = {}
+        for key, value in dictionary.items():
+            if isinstance(dictionary[key], (dict, list)):
+                # If a sub dictionary or a list call this method recursively
+                value = __replace_in_keys(value, s, r)
+            final_dict[key.replace(s, r)] = value
+
+        return final_dict
+
+    # By default return the same object
+    else:
+        return dictionary
+
+
+def _replace_in_keys(
+        config: Mapping[str, Any],
+        key_modifiers: Dict[str, str]) -> Dict[str, Any]:
+    """
+    Replace strings in the keys of a mapping object.
+
+    Parameters
+    ----------
+    config : Mapping
+        The configuration object (or dictionary) to be modified.
+    key_modifiers : dict
+        The dictionary with the replacements.
+
+    Returns
+    -------
+    dict
+        A dictionary with the modified keys.
+
+    """
+    _config = config
+    # Replace longest strings first
+    # - `sorted(..., reverse=True)` takes care of that
+    for key in sorted(key_modifiers.keys(), reverse=True):
```

Comment:
Sorting like this won't exactly have the desired result. E.g. `sorted(["a", "b", "ab"])` results in `["a", "ab", "b"]`.

Reply:
Oops, you are right. I only had `"__"` and `"_"` in mind:
```python3
_replace_in_keys(
    _replace_in_keys(..., key_modifiers={' ': '__', '-': '_'}),
    key_modifiers={'__': ' ', '_': '-'})
```
in mind.

Obviously I have to sort it properly by length!

Reply:
Done in https://github.com/hoedt/upsilonconf/pull/3/commits/40a2eaf7c240d7cbcef254b571db63e662a6e946.